src/main/java/edu/unc/lib/boxc/migration/cdm/services/PermissionsService.java

```diff
@@ -290,27 +291,59 @@ private String getObjectType(String id) {
         List<List<String>> updatedRecords = new ArrayList<>();
-        Set<String> cdmIds = new HashSet<>();
+        Set<String> addedAndUpdatedIds = new HashSet<>();
         String everyoneField = getAssignedRoleValue(options.isStaffOnly(), options.getEveryone());
         String authenticatedField = getAssignedRoleValue(options.isStaffOnly(), options.getAuthenticated());
 
-        // update existing entry
+        // addedAndUpdatedIds: list of all ids that need to be added and updated
+        if (options.getCdmId() != null) {
+            addedAndUpdatedIds.add(options.getCdmId());
+        }
+        if (options.isWithWorks() || options.isWithFiles()) {
+            workAndFileRecords = queryForMappedIds(options);
+            Set<String> workFileIds = workAndFileRecords.stream().map(Map.Entry::getKey).collect(Collectors.toSet());
+            addedAndUpdatedIds.addAll(workFileIds);
+        }
+
+        // update existing entries and add unchanged entries to updatedRecords, then remove updated ids from updateIds
         for (CSVRecord record : previousRecords) {
-            cdmIds.add(record.get(0));
-            if (record.get(0).equals(options.getCdmId())) {
+            if (addedAndUpdatedIds.contains(record.get(0))) {
                 updatedRecords.add(Arrays.asList(record.get(0), record.get(1), everyoneField, authenticatedField));
+                addedAndUpdatedIds.remove(record.get(0));
             } else {
                 updatedRecords.add(Arrays.asList(record.get(0), record.get(1), record.get(2), record.get(3)));
+                addedAndUpdatedIds.remove(record.get(0));
             }
         }
 
-        // add new entry
-        if (!cdmIds.contains(options.getCdmId())) {
+        // add one new entry
+        if (addedAndUpdatedIds.contains(options.getCdmId())) {
             String objectType = getObjectType(options.getCdmId());
             updatedRecords.add(Arrays.asList(options.getCdmId(), objectType, everyoneField, authenticatedField));
         }
 
+        // add new works or files entries
+        for (Map.Entry<String, String> workAndFileRecord : workAndFileRecords) {
+            if (addedAndUpdatedIds.contains(workAndFileRecord.getKey())) {
+                updatedRecords.add(Arrays.asList(workAndFileRecord.getKey(), workAndFileRecord.getValue(),
+                        everyoneField, authenticatedField));
+            }
+        }
+
+        updatedRecords.sort(Comparator.comparing(entry -> entry.get(0)));
+        // move default entry to top if it exists
+        List<String> updatedIds = updatedRecords.stream().map(entry -> entry.get(0)).collect(Collectors.toList());
+        int i = 0;
+        for (String id : updatedIds) {
+            if (id.equals(PermissionsInfo.DEFAULT_ID)) {
+                List<String> defaultEntry = updatedRecords.remove(i);
+                updatedRecords.add(0, defaultEntry);
+            }
+            i++;
+        }
```

Comment:
I think we could do `updatedIds.indexOf(PermissionsInfo.DEFAULT_ID)` instead of the loop